src/main/java/com/crowdin/action/UploadTranslationsAction.java

```diff
@@ -1,71 +1,75 @@
 
-        List<com.crowdin.client.sourcefiles.model.File> files = crowdin.getFiles(branchId);
-        Map<Long, Directory> dirs = crowdin.getDirectories(branchId);
-        Map<String, File> filePaths = CrowdinFileUtil.buildFilePaths(files, dirs);
+            List<com.crowdin.client.sourcefiles.model.File> files = crowdin.getFiles(branchId);
+            Map<Long, Directory> dirs = crowdin.getDirectories(branchId);
+            Map<String, File> filePaths = CrowdinFileUtil.buildFilePaths(files, dirs);
 
-        int uploadedFilesCounter = 0;
+            AtomicInteger uploadedFilesCounter = new AtomicInteger(0);
 
-        for (String src : sourcesList) {
-            VirtualFile source = FileUtil.getSourceFile(root, src);
-            String baseDir = source.getParent().getParent().getPath() + "/";
-            String sourcePath = source.getName();
+            properties.getSourcesWithPatterns().forEach((sourcePattern, translationPattern) -> {
+                List<VirtualFile> sources = FileUtil.getSourceFilesRec(root, sourcePattern);
+                sources.forEach(source -> {
+                    VirtualFile baseDir = FileUtil.getBaseDir(source, sourcePattern);
+                    String sourcePath = source.getName();
 
-            File crowdinSource = filePaths.get(sourcePath);
-            if (crowdinSource == null) {
-                NotificationUtil.showWarningMessage(project, "File '" + (branch != null ? branch + "/" : "") + sourcePath + "' is missing in the project. Run 'Upload' to upload the missing source");
-                continue;
-            }
-            String pattern1 = PlaceholderUtil.replaceFilePlaceholders(PATTERN, sourcePath);
-            for (Language lang : projectLanguages) {
-                String pattern2 = PlaceholderUtil.replaceLanguagePlaceholders(pattern1, lang);
-                java.io.File translationFile = new java.io.File(baseDir + pattern2);
-                if (!translationFile.exists()) {
-                    continue;
-                }
-                crowdin.uploadTranslationFile(translationFile, crowdinSource.getId(), lang.getId());
-                uploadedFilesCounter++;
-            }
+                    File crowdinSource = filePaths.get(sourcePath);
+                    if (crowdinSource == null) {
+                        NotificationUtil.showWarningMessage(project, "File '" + (branch != null ? branch + "/" : "") + sourcePath + "' is missing in the project. Run 'Upload' to upload the missing source");
+                        return;
+                    }
+                    String pattern1 = PlaceholderUtil.replaceFilePlaceholders(translationPattern, sourcePath);
+                    for (Language lang : projectLanguages) {
+                        String pattern2 = PlaceholderUtil.replaceLanguagePlaceholders(pattern1, lang);
+                        java.io.File translationFile = new java.io.File(baseDir.getPath() + "/" + pattern2);
```

Comment:
Use `Paths` for concatenation because file separator might be different depending on OS where our plugin will be running